### ISSUE TITLE
zcatalog QA checks

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#2086]
Suggestions from @stephjuneau regarding coadds and zcatalog creation:

Regardless if we're re-running for DR1 or only for DR2, I suggest adding data QA and sanity checks for e.g. uniqueness of TARGETIDs (when they should be) and presence of missing data (NaN's or sometimes zero's) or unexpected values that should raise a flag (e.g., RA outside 0-360 range)

TO DO: this is not yet implemented in the main code but the VAC code:
- check on uniqueness of TARGETID for a given survey, program for zpix
- checks on uniqueness of {TARGETID, TILEID} for a given survey, program for ztile
(Note: both of these checks fail for negative TARGETIDs in SV)

**Comment by sbailey** [comment]
Implementation notes:

Other than the ZCAT/SPEC_PRIMARY columns, zcatalog creation does not do any calculation; it is only propagating and stacking HDUs from redrock and coadd files.  So QA at the time of zcatalog creation is too late in the sense that the problematic data is already on disk in other files.  At the same time, adding QA while writing the coadds is potentially dangerous for the next prod if it could case a small fraction of coadds to fail.  i.e. we want to add QA at the time the data are first generated (not downstream propagation), but we also want to check it on existing data prior to the next prod.

Suggested approach:
  * write QA that can be run on pre-existing FIBERMAP, EXP_FIBERMAP, and REDSHIFTS HDUs without having to re-generate them
  * run that QA on all of iron and daily since iron to catch any problems
  * fix any problems found
  * add QA to assemble_fibermap (which creates the original FIBERMAP from the raw data) and coadd_fibermap (which generates the coadded FIBERMAP and EXP_FIBERMAP).  Failing QA should raise an exception, since simply reporting an error to the log is likely to be missed.  This will help us quickly catch new problems that arise in the raw data.

Note that there are cases like missing columns in input coordinates files resulting in things like FIBER_RA/DEC not being known, but having the data otherwise ok.  We should only have QA crash when something is both wrong and fixable.

**Comment by weaverba137** [comment]
I believe this must be done, since we're finalizing the zcatalog files now.

**Comment by sbailey** [comment]
Yes and no.  We ran QA checks on the updated iron catalogs, but these have not been ported back into the automated pipeline to run the same checks as part of every prod at the point the data are originally produced (e.g. in assemble_fibermap, as opposed to when combining the final redshift catalogs, at which point it is too late to fix upstream problems).  See the second post in this ticket for concerns about how to implement this and pre-test to make sure we don't accidentally break future prods.
